### PR TITLE
AK: Use C++20 concepts to only allow Userspace wrappers of pointers

### PR DESCRIPTION
**Pull request by emanuele6**
It was a bit odd that you could create a `Userspace<int>` and that `Userspace<int>::ptr()` returned an `int` instead of an `int*`.
    
Let's use C++20 concepts to only allow creating `Userspace` objects with pointer types. :^)

We could also rewrite `Userspace<T>` as something like `UserspacePtr<T>` and make it store `m_ptr` (etc.) as a `T*`, but since we are using `Userspace` to also wrap const-pointers and void-pointers we would need to make separate classes for those (something like `ConstUserspacePtr<T>` and `VoidUserspacePtr`).

The C++20 concepts solution is definitely cleaner.

-----------------------------
I also replaced `#if defined(KERNEL)` with `#ifdef KERNEL` in `Userspace.h`.

### CODE REVIEW

File: AK/Userspace.h [hunk 1]
```diff
@@ -27,10 +27,14 @@
 #pragma once
 
 #include <AK/Types.h>
+#include <type_traits>
```
Comment (awesomekling) [inline AK/Userspace.h:30]: Please don't use STL headers. Instead, use `IsPointer` from AK/StdLibExtras.h :)

Comment (emanuele6) [reply AK/Userspace.h:30]: Oh, ok. I didn't know that existed. I will do it now.

**Comment by emanuele6** [comment]
```log
In file included from /home/travis/build/SerenityOS/serenity/./Kernel/API/Syscall.h:30,
                 from /home/travis/build/SerenityOS/serenity/AK/SharedBuffer.cpp:31:
/home/travis/build/SerenityOS/serenity/./AK/Userspace.h:35:1: error: ‘concept’ does not name a type
 concept PointerTypeName = IsPointer<T>::value;
 ^~~~~~~
/home/travis/build/SerenityOS/serenity/./AK/Userspace.h:35:1: note: ‘concept’ only available with -fconcepts
/home/travis/build/SerenityOS/serenity/./AK/Userspace.h:37:10: error: ‘PointerTypeName’ has not been declared
 template<PointerTypeName T>
          ^~~~~~~~~~~~~~~
/home/travis/build/SerenityOS/serenity/./AK/Userspace.h:54:16: error: expected ‘)’ before ‘ptr’
     Userspace(T ptr)
              ~ ^~~~
                )
/home/travis/build/SerenityOS/serenity/./AK/Userspace.h:59:5: error: ‘T’ does not name a type
     T ptr() const { return m_ptr; }
     ^
/home/travis/build/SerenityOS/serenity/./AK/Userspace.h:66:5: error: ‘T’ does not name a type
     T m_ptr { nullptr };
     ^
```
I can be lagom fine with `gcc`, travis is probably using `clang` to build `lagom` and needs `-fconcepts`.

File: CMakeLists.txt [hunk 2]
```diff
@@ -40,7 +40,7 @@ add_custom_target(check-style
     USES_TERMINAL
 )
 
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wextra -Werror -std=c++2a -fdiagnostics-color=always")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wextra -Werror -std=c++2a -fdiagnostics-color=always -fconcepts")
```
Comment (emanuele6) [inline CMakeLists.txt:43]: Let's see if this works.

Comment (emanuele6) [reply CMakeLists.txt:43]: It worked.

**Pull request merged**